sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -62,7 +62,11 @@ import org.apache.spark.util.Utils
 
 private[sql] object Dataset {
   def apply[T: Encoder](sparkSession: SparkSession, logicalPlan: LogicalPlan): Dataset[T] = {
-    new Dataset(sparkSession, logicalPlan, implicitly[Encoder[T]])
+    val dataset = new Dataset(sparkSession, logicalPlan, implicitly[Encoder[T]])
+    // Eagerly bind the encoder so we verify that the encoder matches the underlying
+    // schema. The user will get an error if this is not the case.
```

Comment:
shall we do the same for `ofRows` below?

Reply:
This is needed because we made `Dataset.deserializer` lazy. The idea there is that we only need to create a `deserializer` (which is expensive) when we materialize data. However for Datasets we also need to make sure that we can actually deserialize an InternalRow into the Dataset type. This does not apply to Dataframes and the `ofRows` method because we are guaranteed that that an InternalRow can be deserialized into a Row.